### PR TITLE
Adm-688:[frontend] [backend] [stub] Split async generate report API into 3 parts 

### PR DESCRIPTION
**Pull request by Zxx-Ivy**
## Summary

...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java [hunk 1]
```diff
@@ -56,7 +59,7 @@ public ResponseEntity<CallbackResponse> generateReport(@RequestBody GenerateRepo
 			.body(CallbackResponse.builder().callbackUrl(callbackUrl).interval(interval).build());
 	}
 
-	@GetMapping("/{dataType}/{filename}")
+	@GetMapping("/reports/{dataType}/{filename}")
 	public InputStreamResource exportCSV(@PathVariable String dataType, @PathVariable String filename) {
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/report/GenerateReportController.java:63]: `dataType` should be a enum

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java [hunk 2]
```diff
@@ -65,15 +68,79 @@ public InputStreamResource exportCSV(@PathVariable String dataType, @PathVariabl
...
 			log.info("Successfully generate Report, report id: {}, reports: {}", reportId, reportResponse);
+			// todo: calculate the time to generate csv file
+			generateReporterService.generateCSVForMetric(reportResponse, reportId);
 			return ResponseEntity.status(HttpStatus.CREATED).body(reportResponse);
 		}
-		return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+		return ResponseEntity.status(HttpStatus.OK).body(reportResponse);
+	}
+
+	@PostMapping("/board-reports")
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/report/GenerateReportController.java:85]: Wrong pattern for the restful design!
POST `reports/{reportType}`
Refactor the rest RESTful API

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 3]
```diff
@@ -245,26 +251,30 @@ public synchronized ReportResponse generateReporter(GenerateReportRequest reques
...
-		saveReporterInHandler(reportResponse, request.getCsvTimeStamp());
 		return reportResponse;
 	}
 
 	private FetchedData fetchOriginalData(GenerateReportRequest request, List<String> lowMetrics) {
 		FetchedData fetchedData = new FetchedData();
 
 		if (lowMetrics.stream().anyMatch(this.kanbanMetrics::contains)) {
+			if (request.getJiraBoardSetting() == null)
+				throw new BadRequestException("Jira board setting is null.");
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:263]: all exception/failed message should be started with `Failed to do ....`

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 4]
```diff
@@ -562,12 +572,87 @@ private void generateCSVForPipeline(GenerateReportRequest request, BuildKiteData
...
+
+	private Boolean getCombinedReadyValue(Boolean previousReadyValue, Boolean newReadyValue) {
+		return previousReadyValue != null || newReadyValue == null ? previousReadyValue : newReadyValue;
+	}
+
+	public void updateMetricsDataReadyInHandler(String timeStamp, List<String> metrics) {
+		MetricsDataReady metricsStatus = getMetricsStatus(metrics, Boolean.TRUE);
+		MetricsDataReady previousMetricsReady = asyncReportRequestHandler.getMetricsDataReady(timeStamp);
+		if (previousMetricsReady == null) {
+			log.error("Unable update metrics data ready through this timestamp.");
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:620]: Update error message with our log convention

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 5]
```diff
@@ -708,8 +799,35 @@ public Boolean deleteExpireCSV(Long currentTimeStamp, File directory) {
 		}
 	}
 
+	// todo: need change this to private after the "/reports" endpoint is deprecated
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:802]: fix this todo

**Review by guzhongren — CHANGES_REQUESTED** [review]
So biiiiiiiiiger PR,
PLZ Split this PR into multiple small PRs

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java (hunk 4 above)
Comment (mikeyangyun) [reply backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:620]: Already update related error message.

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java (hunk 5 above)
Comment (mikeyangyun) [reply backend/src/main/java/heartbeat/service/report/GenerateReporterService.java]: Already checked and fixed.

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java (hunk 3 above)
Comment (mikeyangyun) [reply backend/src/main/java/heartbeat/service/report/GenerateReporterService.java]: Already update exception message which will follow our convention.

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java (hunk 2 above)
Comment (Zxx-Ivy) [reply backend/src/main/java/heartbeat/controller/report/GenerateReportController.java]: done

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java (hunk 1 above)
Comment (Zxx-Ivy) [reply backend/src/main/java/heartbeat/controller/report/GenerateReportController.java]: done

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java [hunk 6]
```diff
@@ -68,12 +52,75 @@ public InputStreamResource exportCSV(@PathVariable String dataType, @PathVariabl
...
+		CompletableFuture.runAsync(() -> {
+			try {
+				ReportResponse reportResponse = generateReporterService
+					.generateReporter(request.convertToReportRequest());
+				generateReporterService.saveReporterInHandler(reportResponse,
+						IdUtil.getBoardReportId(request.getCsvTimeStamp()));
+				generateReporterService.updateMetricsDataReadyInHandler(request.getCsvTimeStamp(),
+						request.getMetrics());
+				log.info(
+						"Generate Board Report successfully, metrics: {}, consider holiday: {}, start time: {}, end time: {}, board-report id: {}",
```
Comment (Lei010) [inline backend/src/main/java/heartbeat/controller/report/GenerateReportController.java:81]: Success do something

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java [hunk 7]
```diff
@@ -68,12 +52,75 @@ public InputStreamResource exportCSV(@PathVariable String dataType, @PathVariabl
...
 			return ResponseEntity.status(HttpStatus.CREATED).body(reportResponse);
 		}
-		return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+		return ResponseEntity.status(HttpStatus.OK).body(reportResponse);
+	}
+
+	@PostMapping("/board")
+	public ResponseEntity<CallbackResponse> generateBoardReport(@RequestBody GenerateBoardReportRequest request) {
+		log.info(
+				"Start to generate Board Report, metrics: {}, consider holiday: {}, start time: {}, end time: {}, board-report id: {}",
```
Comment (Lei010) [inline backend/src/main/java/heartbeat/controller/report/GenerateReportController.java:68]: 1.Board report to lower case
2.need to show request in log

File: backend/src/main/java/heartbeat/controller/report/dto/response/ReportResponse.java [hunk 8]
```diff
@@ -29,6 +29,14 @@ public class ReportResponse {
 
 	private Long exportValidityTime;
 
+	private Boolean boardMetricsReady;
+
+	private Boolean pipelineMetricsReady;
+
+	private Boolean sourceControlMetricsReady;
+
+	private Boolean allMetricsReady;
```
Comment (Lei010) [inline backend/src/main/java/heartbeat/controller/report/dto/response/ReportResponse.java:38]: 1.change to isMetricsReady pattern

File: backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java [hunk 9]
```diff
@@ -40,20 +38,13 @@
...
 import io.micrometer.core.instrument.util.TimeUtils;
 
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.StandardCopyOption;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (Lei010) [inline backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java:47]: weird import

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 10]
```diff
@@ -562,12 +572,87 @@ private void generateCSVForPipeline(GenerateReportRequest request, BuildKiteData
...
 	}
 
-	private void generateCSVForMetric(ReportResponse reportResponse, String csvTimeStamp) {
+	public void generateCSVForMetric(ReportResponse reportResponse, String csvTimeStamp) {
 		csvFileGenerator.convertMetricDataToCSV(reportResponse, csvTimeStamp);
 	}
 
-	private void saveReporterInHandler(ReportResponse reportResponse, String csvTimeStamp) {
-		asyncReportRequestHandler.put(csvTimeStamp, reportResponse);
+	public void saveReporterInHandler(ReportResponse reportResponse, String reportId) {
```
Comment (Lei010) [inline backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:579]: need to change variable name from reportResponse to reportContent

File: backend/src/test/java/heartbeat/controller/report/GenerateReporterControllerTest.java [hunk 11]
```diff
@@ -160,18 +144,90 @@ void shouldReturnWhenExportCsv() throws Exception {
...
+			.deploymentFrequency(DeploymentFrequency.builder()
+				.avgDeploymentFrequency(new AvgDeploymentFrequency("Average", 0.10F))
+				.build())
+			.build();
+
 		ObjectMapper mapper = new ObjectMapper();
-		GenerateReportRequest request = mapper
-			.readValue(new File("src/test/java/heartbeat/controller/report/request.json"), GenerateReportRequest.class);
+		GenerateBoardReportRequest request = mapper.readValue(
+				new File("src/test/java/heartbeat/controller/report/request.json"), GenerateBoardReportRequest.class);
```
Comment (Lei010) [inline backend/src/test/java/heartbeat/controller/report/GenerateReporterControllerTest.java:157]: 1.extract variable: request.json file path and csvtime stamp
2.remove given when then comments in unit test

File: backend/src/test/java/heartbeat/controller/report/GenerateReporterControllerTest.java [hunk 12]
```diff
@@ -160,18 +144,90 @@ void shouldReturnWhenExportCsv() throws Exception {
...
+		GenerateDoraReportRequest request = mapper.readValue(
+				new File("src/test/java/heartbeat/controller/report/request.json"), GenerateDoraReportRequest.class);
 		String currentTimeStamp = "1685010080107";
 		request.setCsvTimeStamp(currentTimeStamp);
 
 		RequestFailedException requestFailedException = new RequestFailedException(402, "Client Error");
-		when(generateReporterService.generateReporter(request)).thenThrow(requestFailedException);
+		when(generateReporterService.generateReporter(request.convertToReportRequest()))
+			.thenThrow(requestFailedException);
+		doNothing().when(generateReporterService).initializeMetricsDataReadyInHandler(any(), any());
```
Comment (Lei010) [inline backend/src/test/java/heartbeat/controller/report/GenerateReporterControllerTest.java:225]: change any to more accurate parameter

File: backend/src/test/java/heartbeat/service/report/CSVFileGeneratorTest.java [hunk 13]
```diff
@@ -15,12 +15,7 @@
 import org.mockito.quality.Strictness;
 import org.springframework.core.io.InputStreamResource;
 
-import java.io.BufferedReader;
-import java.io.File;
-import java.io.FileInputStream;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
+import java.io.*;
```
Comment (Lei010) [inline backend/src/test/java/heartbeat/service/report/CSVFileGeneratorTest.java:18]: weird import

File: backend/src/test/java/heartbeat/handler/AsyncExceptionHandlerTest.java [hunk 14]
```diff
@@ -20,22 +21,25 @@ void shouldDeleteAsyncException() {
...
 	void shouldPutAndGetAsyncException() {
 		long currentTimeMillis = System.currentTimeMillis();
 		String currentTime = Long.toString(currentTimeMillis);
-		asyncExceptionHandler.put(currentTime, new UnauthorizedException("test"));
+		String boardReportId = IdUtil.getBoardReportId(currentTime);
+		asyncExceptionHandler.put(boardReportId, new UnauthorizedException("test"));
 
-		BaseException baseException = asyncExceptionHandler.get(currentTime);
+		BaseException baseException = asyncExceptionHandler.get(boardReportId);
 		assertEquals(401, baseException.getStatus());
```
Comment (Lei010) [inline backend/src/test/java/heartbeat/handler/AsyncExceptionHandlerTest.java:43]: try using http status code

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 15]
```diff
@@ -245,26 +251,30 @@ public synchronized ReportResponse generateReporter(GenerateReportRequest reques
...
 		if (lowMetrics.stream().anyMatch(this.codebaseMetrics::contains)) {
+			if (request.getCodebaseSetting() == null)
+				throw new BadRequestException("Failed to fetch Github info due to code base setting is null.");
 			BuildKiteData buildKiteData = fetchGithubData(request);
 			fetchedData.setBuildKiteData(buildKiteData);
 		}
 
 		if (lowMetrics.stream().anyMatch(this.buildKiteMetrics::contains)) {
+			if (request.getBuildKiteSetting() == null)
+				throw new BadRequestException("Failed to fetch BuildKite info due toBuildKite setting is null.");
```
Comment (Lei010) [inline backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:277]: due to

**Review by guzhongren — APPROVED** [review]

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java (hunk 6 above)
Comment (sqsq5566) [reply backend/src/main/java/heartbeat/controller/report/GenerateReportController.java]: done

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java (hunk 7 above)
Comment (sqsq5566) [reply backend/src/main/java/heartbeat/controller/report/GenerateReportController.java]: lower case is done

File: backend/src/main/java/heartbeat/controller/report/dto/response/ReportResponse.java (hunk 8 above)
Comment (Lei010) [reply backend/src/main/java/heartbeat/controller/report/dto/response/ReportResponse.java:38]: done

File: backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java (hunk 9 above)
Comment (Lei010) [reply backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java:47]: done

File: backend/src/test/java/heartbeat/service/report/CSVFileGeneratorTest.java (hunk 13 above)
Comment (Lei010) [reply backend/src/test/java/heartbeat/service/report/CSVFileGeneratorTest.java:18]: done

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java (hunk 10 above)
Comment (Lei010) [reply backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:579]: done

File: backend/src/test/java/heartbeat/handler/AsyncExceptionHandlerTest.java (hunk 14 above)
Comment (Lei010) [reply backend/src/test/java/heartbeat/handler/AsyncExceptionHandlerTest.java:43]: done

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java (hunk 15 above)
Comment (Lei010) [reply backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:277]: done

File: backend/src/test/java/heartbeat/controller/report/GenerateReporterControllerTest.java (hunk 11 above)
Comment (Lei010) [reply backend/src/test/java/heartbeat/controller/report/GenerateReporterControllerTest.java:157]: done

**Pull request merged**